check_pgactivity

```diff
@@ -5600,6 +5604,186 @@ sub check_oldest_idlexact {
+            UNION ALL SELECT NULL, NULL
+          ) sql(transaction, datname)
+          -- we use this JOIN condition to make sure that we'll always have a
+          -- full record for all (connectable) databases
+          JOIN pg_database d ON d.datname = coalesce(sql.datname, d.datname)
+          WHERE d.datallowconn
+        )
+        SELECT kind, datname, age, xmin FROM ordered
+        WHERE rownum = 1
+        },
+        # backend_xmin and backend_xid added to pg_stat_activity,
+        # backend_xmin added to pg_stat_replication,
+        # replication slots introduced
+        $PG_VERSION_94 => q{
+        WITH raw AS (
+          -- regular backends
+          SELECT
+          CASE WHEN xact_start = query_start
+            THEN 'query'
+            ELSE
+              CASE WHEN state = 'idle in transaction'
+                THEN 'idle_xact'
+                ELSE 'active_xact'
+              END
+          END AS kind,
+          datname,
+          coalesce(backend_xmin, backend_xid) AS xmin
+          FROM pg_stat_activity
+          -- exclude ourselves, as a blocked xmin in another database would be
+          -- exposed in the database we're connecting too, which may otherwise
+          -- not have the same xmin
+          WHERE pid != pg_backend_pid()
+          UNION ALL (
+            -- 2PC
+            SELECT '2pc' AS kind,
+            database AS datname,
+            transaction AS xmin
+            FROM  pg_prepared_xacts
+          ) UNION ALL (
+            -- replication slots
+            SELECT 'repslot' AS kind,
+            database AS datname,
+            xmin AS xmin
+            FROM  pg_replication_slots
+          ) UNION ALL (
+            -- walsenders
+            SELECT 'walsender' AS kind,
+            NULL AS datname,
+            backend_xmin AS xmin
+            FROM pg_stat_replication
+          )
+        ),
+        ordered AS (
+          SELECT f.kind, d.datname,
+          -- xid type doesn't have range operators as the value will wraparound.
+          -- Instead, rely on age() function and row_number() window function
+          -- to get the oldest xid found.
+          row_number() OVER (
+            PARTITION BY f.kind, d.datname
+            ORDER BY age(xmin) DESC NULLS LAST
+          ) rownum, age(xmin) AS age, xmin
+          FROM raw
+          RIGHT JOIN (VALUES
+            ( 'query'       ),
+            ( 'idle_xact'   ),
+            ( 'active_xact' ),
+            ( '2pc'         ),
+            ( 'repslot'     ),
+            ( 'walsender'   )
+          ) f(kind) ON raw.kind = f.kind
+          -- we use this JOIN condition to make sure that we'll always have a
+          -- full record for all (connectable) databases
+          JOIN pg_database d ON d.datname = coalesce(raw.datname, d.datname)
+          WHERE d.datallowconn
+        )
+        SELECT kind, datname, age, xmin
+        FROM ordered
+        WHERE rownum = 1
+        }
```

Comment:
This query reports as many walsender than existing database, which is wrong.

Reply:
This returns as many physical walsender as connectable database, which is exactly what is expected.  It simplifies knowing what is the xmin horizon on a given database without checking the possible "global" limits.

Reply:
physical walsender are global, they are not related to one single database. So I don't understand your answer. It's still confusing

Reply:
I understood why it has been done this way after discussing. Closing this conversation